crossdock/server/server.py

```diff
@@ -130,79 +168,88 @@ def handle_trace_request(self, http_request, trace_request, span_handler, respon
         if span_handler:
             span_handler(span)
 
-        trace_id = "%x" % span.trace_id
-        observed_span = ObservedSpan(
-            trace_id, span.is_sampled(),
-            span.get_baggage_item(constants.baggage_key))
-
-        tr = TraceResponse(span=observed_span)
+        with tchannel.context_provider.span_in_context(span):
+            trace_id = "%x" % span.trace_id
+            observed_span = ObservedSpan(
+                trace_id, span.is_sampled(),
+                span.get_baggage_item(constants.baggage_key))
 
-        if trace_request and trace_request.downstream is not None:
-            self.call_downstream(span, trace_request, tr, response_writer)
-        else:
-            response_writer.write(serializer.obj_to_json(tr))
-            response_writer.finish()
+            tr = TraceResponse(span=observed_span)
 
-    def call_downstream(self, span, trace_request, trace_response, response_writer):
-        if trace_request.downstream.transport == Transport.HTTP:
-            self.call_downstream_http(span, trace_request, trace_response, response_writer)
-        elif trace_request.downstream.transport == Transport.TCHANNEL:
-            if tchannel_supported:
-                self.real_call_downstream_tchannel(span, trace_request, trace_response,
-                                                   response_writer)
+            if trace_request and trace_request.downstream is not None:
+                call_downstream(span, trace_request, tr, response_writer)
```

Comment:
call_downstream is a coroutine, how are we calling it here as a plain function? And shouldn't the endpoint handling methods also be coroutines?